### ISSUE TITLE
Unusual cloud rendering bug

### ISSUE THREAD

**Issue by BLOKBUSTR** [Rivmun/SuperFancyCloudsRefabricated#49]
This just seemingly happened out of the blue, so I cannot provide steps to replicate this bug, though I believe this could've occurred because I was playing on the world for too long. At some point in my world I noticed that the clouds formed a large sort of wall in a sylinder shape around my position. I assume this may be related to the render distance. If it helps provide any pointers to a solution, I have "Cloud Render Distance Fit to View" disabled. I've re-enabled this for the time being to see whether this bug would reoccur.

![2023-04-19_19 17 02](https://user-images.githubusercontent.com/118950712/233241519-2b202957-9db6-421a-ac4c-3124f419a5f4.png)

**Comment by Rivmun** [comment]
Seems you enabled "Biome Detect From Chunk" in Dynamic config tab, right? How about "- Use Loaded Chunk"?
I guess it causing by density value error when calculate outer clouds
Try to enable "Debug" option when it occurs again and see what log says.

**Comment by BLOKBUSTR** [comment]
It's possible that I might've had "Biome Detect From Chunk" enabled and "- Use Loaded Chunk" disabled at that moment, though I don't exactly remember.

I have recorded almost three minutes of myself testing what would happen when I changed these settings and uploaded the video and the log itself to a Dropbox folder. The events recorded in the video start at the log timestamp 20:47:52. It seems the strangest results come from when both options are disabled.
https://www.dropbox.com/sh/8y7uzpq7v18oyvg/AADRN5jls0w6X1TP4HOYtgzsa?dl=0

I am using version 1.7.0 of SFCR, as well as:
- Sodium 1.19.2-0.4.4+build.18
- Indium 1.0.9

I think I might also want to mention straightaway that I'm using the Terralith datapack, so it's *possible* that it might be influencing the behavior of SFCR in some way with all its different biomes, though I cannot be sure. I really hope this helps; I'm completely free tomorrow, so I can provide any further information or testing if needed.

**Comment by Rivmun** [comment]
I saw this video and log. It's very helpful.

But I have no idea how this issue happens. I try to analyze and found something strange:
- In video, weather is clear stably, but density value affect by weather is continuing increase (Shown in log with "wd"), near to density in thunder (1.8). It's presenting the weather detection is out of function. How?
- There is shouldn't a hole in the cloud layer when disabled "Biome Detect From Chunk". It should be flat & full because the weather density is too high.
  - I'm tested with Terralith and found that it's breaks SFCR's "Terrain Dodge" function. The hole maybe causing by it.
- Even the weather density is high, cloud shouldn't gen so solidly.
  - Technically, the sample result is a number range around [-1, 2]. When weather density (wd) is 1.8 and biome density (bd) is 0.5, the result of gen gate is -0.125. Clouds still has around 33% chance not to gen (when sample < gen gate). It shouldn't be continuously.

I decide to fix the "Terrain Dodge" to compatible with Terralith first. I'll mention you and paste patch file to here when it done.

**Comment by Rivmun** [comment]
I'm fixed the terrain dodge with Terralith and add some debug output in weather calculation. Download this dev build in Artifacts and try.
https://github.com/Rivmun/SuperFancyCloudsRefabricated/actions/runs/4771052503

**Comment by BLOKBUSTR** [comment]
I tested the Artifact on a clean installation with very minimal mods, and the weather density values appear to be far more consistent. I have not noticed any abnormal cloud shapes or holes either during my testing.

I uploaded the latest log to the same Dropbox folder above and renamed the old log and video accordingly. In the game chat I did mention another recording, but I decided it wasn't really worth uploading as nothing anomalous happened anyway. I'll still keep it on my machine for now just in case.

**Comment by Rivmun** [comment]
Thanks for your logs. I find some dubious problem that may causing weather detection failure:

> [13:13:05] [Server thread/INFO]: thndTime: 13431, rainTime: 87813, clearTime: 0
> [13:13:05] [Server thread/INFO]: currentWeather: CLEAR, nextWeather: THUNDER

If the weather detect time is default (10s), nextWeather is wrong in here. It should be CLEAR when thndTime is bigger than 200 (ticks).

> [13:13:35] [Server thread/INFO]: thndTime: 12831, rainTime: 87213, clearTime: 0
> [13:13:35] [Server thread/INFO]: currentWeather: CLEAR, nextWeather: THUNDER
> [13:13:36] [Render thread/INFO]: wc: false, bc: true, wd: 0.5, bd: 0.69609356

If currentWeather is not equal to nextWeather, weather changing (wc) shouldn't be _false_.

Density value was correct is a lucky thing. 😂 Wait a few moments, I'll dig it out why.

**Comment by Rivmun** [comment]
> If currentWeather is not equal to nextWeather, weather changing (wc) shouldn't be false.

Fixed in 2e82e3be153f0a20933b0317c39b58a4fd6929ab. It's causing by weatherChanging has been override when client ticking the weather. I should disable it when there is a server (or integrated server if in singleplay). You can download the latest build in Actions tab.

> If the weather detect time is default (10s), nextWeather is wrong in here. It should be CLEAR when thndTime is bigger than 200 (ticks).

The question is "why nextWeather always returns THUNDER". It's complex to explain what happen.
nextWeather is assigned value from asking server "what weather now". After I met the same issue as your question and excluding options that may have errors in the assigned method, I found the server is really answer that "is thunder now", even there is no rain or darker sky on client's gameplay.
It's very weird and I don't know why, because the answer is coming from Minecraft Server. Unfortunately, I don't have the capacity to debug for Minecraft itself, so I cannot fix it.

In summary, I can only suggest that try to close the "weather pre-detect" by set it to 0 when the "clouds wall" coming again. It may avoid that.

**Comment by Rivmun** [comment]
The next weather change after using "/weather clear" will always be a thunderstorm
Quote from https://bugs.mojang.com/browse/MC-108806
